### ISSUE TITLE
Dependancy on click should be optional.

### ISSUE THREAD

**Issue by rjbez17** [jfinkhaeuser/prance#27]
## Expected Behaviour
I can use this as a library only with no dependency on click

## Actual Behaviour
My CLI (which uses Click 7.0) is unable to use this as a library as there is a hard dependency for Click 6.7. It be nice if this could be used as a library only without any CLI specific dependencies (and not requiring a fork).

@jfinkhaeuser

**Comment by jfinkhaeuser** [comment]
Yup, that makes a lot of sense.

**Comment by rjbez17** [comment]
I'm happy to submit a PR if you provide some feedback on how we want it implemented. There are a few approaches we could take here.

**Comment by jfinkhaeuser** [comment]
Well, making the click dependency an extra_require arg is simple enough. I'd want requirements.txt to default to including the extra. And the cli entry point in setup.py needs to be tagged appropriately. That's around 95% of it.

The remaining 5% are annoying - unfortunately, the flex dependency does not like click 7, and getting rid of that will be more involved.

**Comment by rjbez17** [comment]
Flex is no longer being maintained, so we may want to find an alternative anyway: 

https://github.com/pipermerriam/flex

Another option is to pull the library specific items into a separate repo (prancer maybe?) and have prance (the CLI for prancer) have the library as a dependency. That avoids a lot of build/install time complications. If not correctly it shouldn't break any current users, but allow for a separation between the CLI and Library.

My specific usecase is converting 2.0 -> 3.0 and resolving references. I could write it myself but figured this would be less maintenance ;)

**Comment by jfinkhaeuser** [comment]
Let me stew on this for a day or two. Things have been hectic, and I'm in a different project headspace at the moment. I very much like flex because with the other two, the hackaround due to JSON Schema usage and the performance both are nothing to write home about.

I'm leaning towards making flex optional, then there's a path forward. Meh.

**Comment by jfinkhaeuser** [comment]
So, I took a while (was busy), but I'm going to be dealing with this today.

**Comment by jfinkhaeuser** [comment]
So what's currently in master, and what it seems I will tag as 0.14 should do the trick:

Click 7, no flex:
```bash
$ pip install prance[cli]
```

No click, but flex. For CLI functionality, install click 6.7 separately:
```bash
$ pip install prance[flex]
```

That should give you what you want, and is backwards compatible.

**Comment by rjbez17** [comment]
I'm still seeing a click dependency in a new virtualenv:

```
[prancetest] -----@----- Feb 08,2019 08:11:41:~/projects/prancetest 
$ pip install prance[flex]
Collecting prance[flex]
  Downloading https://files.pythonhosted.org/packages/96/37/564b9944738f3caceaaa85a802f57078ce1341e638cd9f9e333ae11091e5/prance-0.14.0-py2.py3-none-any.whl
Collecting PyYAML>4,>=4.2b4 (from prance[flex])
  Downloading https://files.pythonhosted.org/packages/a8/c6/a8d1555e795dbd0375c3c93b576ca13bbf139db51ea604afa19a2c35fc03/PyYAML-4.2b4.tar.gz (262kB)
    100% |████████████████████████████████| 266kB 2.6MB/s 
Collecting six~=1.12 (from prance[flex])
  Using cached https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Collecting requests~=2.21 (from prance[flex])
  Using cached https://files.pythonhosted.org/packages/7d/e3/20f3d364d6c8e5d2353c72a67778eb189176f08e873c9900e10c0287b84b/requests-2.21.0-py2.py3-none-any.whl
Collecting chardet~=3.0 (from prance[flex])
  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
Collecting semver~=2.8 (from prance[flex])
  Using cached https://files.pythonhosted.org/packages/21/18/a0de8cda637ba3efee1b3617ded00601507ce15bd70a39399740e0fd415f/semver-2.8.1-py2.py3-none-any.whl
Collecting flex; extra == "flex" (from prance[flex])
  Using cached https://files.pythonhosted.org/packages/8d/3d/7075e8a5844e485d80972f4f6e9aef287f5b898312524f58d96a203b794d/flex-6.13.2-py3-none-any.whl
Collecting idna<2.9,>=2.5 (from requests~=2.21->prance[flex])
  Using cached https://files.pythonhosted.org/packages/14/2c/cd551d81dbe15200be1cf41cd03869a46fe7226e7450af7a6545bfc474c9/idna-2.8-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests~=2.21->prance[flex])
  Using cached https://files.pythonhosted.org/packages/9f/e0/accfc1b56b57e9750eba272e24c4dddeac86852c2bebd1236674d7887e8a/certifi-2018.11.29-py2.py3-none-any.whl
Collecting urllib3<1.25,>=1.21.1 (from requests~=2.21->prance[flex])
  Using cached https://files.pythonhosted.org/packages/62/00/ee1d7de624db8ba7090d1226aebefab96a2c71cd5cfa7629d6ad3f61b79e/urllib3-1.24.1-py2.py3-none-any.whl
Collecting validate-email<2,>=1.2 (from flex; extra == "flex"->prance[flex])
Collecting strict-rfc3339<1,>=0.7 (from flex; extra == "flex"->prance[flex])
Collecting jsonpointer<2,>=1.7 (from flex; extra == "flex"->prance[flex])
  Using cached https://files.pythonhosted.org/packages/7b/f4/90449441c1c011a733ee85739a680f2442eda2130e6a88f51804660a0289/jsonpointer-1.14-py2.py3-none-any.whl
Collecting rfc3987<2,>=1.3.4 (from flex; extra == "flex"->prance[flex])
  Using cached https://files.pythonhosted.org/packages/65/d4/f7407c3d15d5ac779c3dd34fbbc6ea2090f77bd7dd12f207ccf881551208/rfc3987-1.3.8-py2.py3-none-any.whl
Collecting click<7,>=3.3 (from flex; extra == "flex"->prance[flex])
  Using cached https://files.pythonhosted.org/packages/34/c1/8806f99713ddb993c5366c362b2f908f18269f8d792aff1abfd700775a77/click-6.7-py2.py3-none-any.whl
Building wheels for collected packages: PyYAML
  Building wheel for PyYAML (setup.py) ... done
  Stored in directory: /Users/------/Library/Caches/pip/wheels/d2/ab/e6/df7b6c9e3b43272fe0164d6a012e4e02e63baaddae4ca6a0d0
Successfully built PyYAML
flex 6.13.2 has requirement PyYAML<4,>=3.11, but you'll have pyyaml 4.2b4 which is incompatible.
Installing collected packages: PyYAML, six, idna, chardet, certifi, urllib3, requests, semver, validate-email, strict-rfc3339, jsonpointer, rfc3987, click, flex, prance
Successfully installed PyYAML-4.2b4 certifi-2018.11.29 chardet-3.0.4 click-6.7 flex-6.13.2 idna-2.8 jsonpointer-1.14 prance-0.14.0 requests-2.21.0 rfc3987-1.3.8 semver-2.8.1 six-1.12.0 strict-rfc3339-0.7 urllib3-1.24.1 validate-email-1.3
[prancetest] ----@---- Feb 08,2019 08:16:43:~/projects/prancetest 
$ pip list |grep click
click          6.7   
```

**Comment by jfinkhaeuser** [comment]
I see, and it's weird.

**Comment by jfinkhaeuser** [comment]
Ah. That part is actually fine; flex itself has a dependency on click. So you're getting the CLI support "for free" with flex.

What's not so great is that there's a conflict with the pyyaml version flex depends on, and the one deemed safe (there was a security warning).

I'll fix that in a 0.14.1 tag, but the TL;DR is that flex is probably no longer a good idea. I'll document it accordingly.

**Comment by jfinkhaeuser** [comment]
The situation with PyYAML is a bit annoying at the moment because of how distutils handle dependencies. Long story short, I made a few minor changes, but those were more cosmetic.

Specifying flex will:
- install click 6.7, because click<7 is a dependency of flex
- issue an error, because we're now looking for flex >=4.2b2, while flex wants <4

Still, flex works just fine. And since we're using `YAML.safe_load`, it shouldn't even matter if we're using an older or newer version...
